WebUI: Implement adjustable dynamic table columns

buinsky opened the pull request:
![columns](https://cloud.githubusercontent.com/assets/9297293/16928116/ab08ab6e-4d39-11e6-9e8d-bed49051ec64.png)

buinsky commented:
@qbittorrent/webdev @sledgehammer999 @glassez Please review the PR.

ngosang commented:
Great work!!!
Before an exhaustive review, my concern is you added more columns to transfer list and you are sending all columns data even if the columns are disabled (hidden).
Could you change this function to return only enabled columns data? I think you will need to add another parameter with the list of columns, comma separated or something like that...
https://github.com/qbittorrent/qBittorrent/blob/master/src/webui/btjson.cpp#L362

buinsky commented:
@ngosang I just tested WebUI with 18 downloading torrents average maindate request size is 0.9 KB (3,2 MB per hour). Do you think this is too much?

glassez commented on src/webui/btjson.cpp:
```diff
@@ -427,6 +456,11 @@ QByteArray btjson::getSyncTorrentPeersData(int acceptedResponseId, QString hash,
         peer[KEY_PEER_FLAGS_DESCRIPTION] = pi.flagsDescription();
         peer[KEY_PEER_RELEVANCE] = pi.relevance();
+
+        QStringList downloadingFiles(torrent->info().filesForPiece(pi.downloadingPieceIndex()));
```
Please don't produce redundant variables.

glassez commented:
>Could you change this function to return only enabled columns data? I think you will need to add another parameter with the list of columns, comma separated or something like that...

I think it should not be such. IMO, we will reduce the response size by the same size increase request.

glassez commented:
>I just tested WebUI with 18 downloading torrents. Average  maindate  request size is 0.9 KB (2.1 MB per hour). This is not too much.

Agree. And I think it shouldn't increase much even in case of 1000 torrents since we transfer modified data only (except the first response of course).

ngosang commented:
@buinsky I tested it. Some improvements.
![](https://i.imgur.com/yKsSBdl.png)
Some items like "Session download" are broken several lines. Could you add http://www.w3schools.com/cssref/pr_text_white-space.asp or a hard space to prevent this?

![](https://i.imgur.com/TX1ZdER.png)
Could you add a vertical separator between columns? Now it's hard to see how to resize the column width. Just like in the QtUI
![](https://i.imgur.com/K2kHXtl.png)

@sledgehammer999 I tested it and I saw the code, it's ready to merge.

buinsky commented:
@ngosang Done.

ngosang commented:
@sledgehammer999 ready to merge

sledgehammer999 commented:
Some issues without looking at C++ code:
1. Commit names should be: "Prevent text wrapping in **menus**" and "Make too tall **menus** scrollable"
2. It seems there is max size in resizable columns. Try a torrent with a very long name.
3. Resizing the "Done" column doesn't make the progress bar expand. If this is a limitation of the underlying CSS then I can merge it as is.
4. In the menu to select columns the thickness of the scrollbar may vary. In Windows 7 it partially covers one item. See picture.
![untitled](https://cloud.githubusercontent.com/assets/273315/20075231/1509109e-a53c-11e6-8189-fc132c2ca3ea.jpg)

buinsky commented:
@sledgehammer999 

> 1. Commit names should be: "Prevent text wrapping in menus" and "Make too tall menus scrollable"

Fixed.

> 2. It seems there is max size in resizable columns. Try a torrent with a very long name.

I can't reproduce it. I can resize any column to any width.

> 3. Resizing the "Done" column doesn't make the progress bar expand.

Fixed.

> 4. In the menu to select columns the thickness of the scrollbar may vary. In Windows 7 it partially covers one item. See picture.

Fixed.

sledgehammer999 commented:
@buinsky can you resolve the merge conflict? I did it locally but I am not sure it is done correctly.
Also I have noticed 2 more bugs:
1. The first 2 columns are the queue number column the column holding the torrent-status icon. Both can be resized freely. Is this wise? However, the name column cannot be resized for some reason.
2. The "Done" column has a resizable progressbar, but the interface kinda freezes when you drag to resize. It isn't snappy.


Again if I try to drag a column to widen it, it stops after a while. I think the max width is ~300px. I use Firefox 50.1.0 on Windows 7.
The other issues seem to be solved.

buinsky commented:
@sledgehammer999 

Bag with 300px limit of column width is fixed. Also unnessesary updates of progress indicator are removed.

The queue number column and the column holding the torrent-status icon in GUI are resizable. I can make them a fixed size in WebUI if you want.

sledgehammer999 commented:
The other issues are fixed. Progressbar seems more responsive now, but still if you drag back n forth it has noticeable freezes. In any case, I am going to merge this as it is. If you can optimize the progressbar resizing more, then do it in a PR. You have put a lot of work into this and waited a lot. I shouldn't delay this anymore.
Thank you.
